### PR TITLE
Implement Asset#{username,password,host,hostname,path,port,protocol} getters/setters

### PR DESCRIPTION
**Pull request by papandreou**
... and a `rootPath` getter/setter as discussed offline :)

Wrt. `rootPath` I'm not really sure if it's a good name for the property -- maybe it should be `rootPathname` to match `Asset#pathname`).

Also, I'm unsure if it should include the file name. Maybe we need an additional property that doesn't?

### CODE REVIEW

**Comment by papandreou** [comment]
An interesting idea might be to just replace `pathname` with `rootPath` as it is now. I don't think you'll ever want to match against the part of the path that's outside of `assetGraph.root`?

**Comment by Munter** [comment]
That sounds bout right. The only thing I've ever had to do with the non-assetgraph-root part of the url has been to work around it by regexing around it with wildcards. It has only been an annoyance

**Comment by papandreou** [comment]
Fixed in https://github.com/assetgraph/assetgraph/commit/a8af0517de17f5085ad427c7205545e254d37f16

**Comment by Munter** [comment]
Should these setters be called in the constructor as well?

Does it make sense to say:

```js
assetGraph.addAsset({
  url: 'https://foo.com/bar',
  username: 'baz',
  fragment: '#hello'
})
```

**Comment by papandreou** [comment]
Hmm, interesting. Yeah, I think that does make sense (except that `fragment` is a property of incoming relations). We already support something similar for `baseName`, `fileName`, and `extension` via https://github.com/assetgraph/assetgraph/blob/8d81b6b49e86aed886a4b415aefe77a968b532b1/lib/assets/Asset.js#L88-L112

And we already make up a url when a non-inline asset doesn't already have one: https://github.com/assetgraph/assetgraph/blob/8d81b6b49e86aed886a4b415aefe77a968b532b1/lib/AssetGraph.js#L204-L207

And that actually does work with eg. `baseName`:
* https://github.com/assetgraph/assetgraph/blob/8d81b6b49e86aed886a4b415aefe77a968b532b1/lib/transforms/bundleSystemJs.js#L916
* https://github.com/assetgraph/assetgraph/blob/8d81b6b49e86aed886a4b415aefe77a968b532b1/lib/transforms/bundleRelations.js#L128

We should create a bunch of test cases and make sure that these new properties support the same feature set.

**Comment by papandreou** [comment]
I've added a bunch of tests now. It seems like it's a bit harder than I imagined to get the properties to DTRT with the "default" url for non-inline assets than are `addAsset`ed without a `url` property.

Also, `pathname` is a bit awkward because it's the entire path, so it overlaps with `fileName`, `baseName`, and `extension`, so they don't compose well together. For instance, I'm not sure what this is supposed to do:

```js
assetGraph.addAsset({
  url: 'https://example.com/',
  pathname: '/yadda/',
  fileName: 'foo.js'
});
```

It'd be easier if `pathname` didn't include the filename, or if there were an alternative property we could use that didn't so we could prohibit combining `pathname` with those other ones.


Finally -- I think we should find a way to fix the inconsistent casing of `Asset#fileName`, `Asset.baseName`, and `Asset#pathname`.

File: lib/assets/Asset.js [hunk 1]
```diff
@@ -479,6 +487,38 @@ class Asset extends EventEmitter {
...
+        }
+        const url = this.url;
+        if (url) {
+            let value;
+            if (url.startsWith(this.assetGraph.root)) {
+                value = url.substr(this.assetGraph.root.length - 1);
+            } else {
+                value = new urlModule.URL(url).pathname;
+            }
+            return value.replace(/\/+[^/]+$/, '/') || '/';
```
[the remark this answers is not in the document]
Comment (papandreou) [reply lib/assets/Asset.js]: Comment from gitter:
maybe path is actually a misnomer. It's actually the "path" part of the url, so it's `/` delimited and unsafe chars must be encoded.
I really doubt there'll be problems on Windows, but the name might be confusing.

It's consistent with the `pathname` property here, which is also url encoded, though:

```js
> new url.URL('file:///foo/bar/qu ux/bazæ')
URL {
  href: 'file:///foo/bar/qu%20ux/baz%C3%A6',
  origin: 'null',
  protocol: 'file:',
  username: '',
  password: '',
  host: '',
  hostname: '',
  port: '',
  pathname: '/foo/bar/qu%20ux/baz%C3%A6',
  search: '',
  searchParams: URLSearchParams {},
  hash: '' }
```

**Pull request merged**